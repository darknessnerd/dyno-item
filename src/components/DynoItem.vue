<template>
  <div
    data-test="root"
    ref="root"
    :style="style"
    :class="[{
      [classNameActive]: enabled,
      [classNameDragging]: dragging,
      [classNameDraggable]: draggable,
      [classNameResizing]: resizing,
      [classNameResizable]: resizable,
    }, className]"
    @mousedown="elementMouseDown"
    @touchstart="elementTouchDown">
    <div
      v-for="handle in actualHandles"
      :key="handle"
      :class="[classNameHandle, classNameHandle + '-' + handle]"
      :style="handleStyle"
      :data-test="handle"
      @mousedown.stop.prevent="handleDown(handle, $event)"
      @touchstart.stop.prevent="handleTouchDown(handle, $event)">
      <slot :name="handle"></slot>
    </div>
    <div ref="content" :style="contentStyle">
      <slot></slot>
    </div>
  </div>
</template>
<script>

import {
  ref, computed, onMounted, onBeforeUnmount, watch, reactive,
} from 'vue';
import events from '@/utils/events';
import useDrag from '@/hooks/drag';
import useTransform from '@/hooks/transform';
import useResize from '@/hooks/resize';
import {
  addEvent,
  removeEvent,
  getBoundSize, getParentSize,
} from '@/utils/dom';

const userSelectNone = {
  userSelect: 'none',
  MozUserSelect: 'none',
  WebkitUserSelect: 'none',
  MsUserSelect: 'none',
};
const userSelectAuto = {
  userSelect: 'auto',
  MozUserSelect: 'auto',
  WebkitUserSelect: 'auto',
  MsUserSelect: 'auto',
};

export default {
  name: 'dyno-item',
  emits: [
    'activated',
    'touchstart',
    'dragging',
    'drag-stop',
    'deactivated',
    'resizing',
    'resize-stop',
  ],
  props: {
    // Component controls
    /**
     * Active state for the component, this props when change emit activated and deactivated event
     */
    active: {
      type: Boolean,
      default: false,
    },
    /**
     * Draggable state for the component.
     */
    draggable: {
      type: Boolean,
      default: true,
    },
    /**
     * UserSelect state for the component.
     */
    disableUserSelect: {
      type: Boolean,
      default: true,
    },
    /**
     * Keep the component ever active
     */
    preventDeactivation: {
      type: Boolean,
      default: false,
    },
    enableNativeDrag: {
      type: Boolean,
      default: false,
    },
    lockAspectRatio: {
      type: Boolean,
      default: false,
    },
    parent: {
      type: Boolean,
      default: false,
    },
    resizable: {
      type: Boolean,
      default: true,
    },
    className: {
      type: String,
      default: 'vdr',
    },
    classNameDraggable: {
      type: String,
      default: 'draggable',
    },
    classNameActive: {
      type: String,
      default: 'active',
    },
    classNameDragging: {
      type: String,
      default: 'dragging',
    },
    classNameResizing: {
      type: String,
      default: 'resizing',
    },
    dragHandle: {
      type: String,
      default: null,
    },
    classNameHandle: {
      type: String,
      default: 'handle',
    },
    dragCancel: {
      type: String,
      default: null,
    },
    onResizeStart: {
      type: Function,
      default: () => true,
    },
    classNameResizable: {
      type: String,
      default: 'resizable',
    },
    onResize: {
      type: Function,
      default: () => true,
    },
    scale: {
      type: [Number, Array],
      default: 1,
      validator: (val) => {
        if (typeof val === 'number') {
          return val > 0;
        }
        return val.length === 2 && val[0] > 0 && val[1] > 0;
      },
    },
    onDragStart: {
      type: Function,
      default: () => true,
    },
    onDrag: {
      type: Function,
      default: () => true,
    },
    w: {
      type: [Number, String],
      default: 200,
      validator: (val) => {
        if (typeof val === 'number') {
          return val > 0;
        }
        return val === 'auto';
      },
    },
    h: {
      type: [Number, String],
      default: 200,
      validator: (val) => {
        if (typeof val === 'number') {
          return val > 0;
        }
        return val === 'auto';
      },
    },
    minWidth: {
      type: Number,
      default: 0,
      validator: (val) => val >= 0,
    },
    minHeight: {
      type: Number,
      default: 0,
      validator: (val) => val >= 0,
    },
    maxWidth: {
      type: Number,
      default: null,
      validator: (val) => val >= 0,
    },
    maxHeight: {
      type: Number,
      default: null,
      validator: (val) => val >= 0,
    },
    x: {
      type: Number,
      default: 0,
    },
    y: {
      type: Number,
      default: 0,
    },
    z: {
      type: [String, Number],
      default: 'auto',
      validator: (val) => (typeof val === 'string' ? val === 'auto' : val >= 0),
    },
    handles: {
      type: Array,
      default: () => ['tl', 'tm', 'tr', 'mr', 'br', 'bm', 'bl', 'ml'],
      validator: (val) => {
        const s = new Set(['tl', 'tm', 'tr', 'mr', 'br', 'bm', 'bl', 'ml']);
        return new Set(val.filter((h) => s.has(h))).size === val.length;
      },
    },
    axis: {
      type: String,
      default: 'both',
      validator: (val) => ['x', 'y', 'both'].includes(val),
    },
    grid: {
      type: Array,
      default: () => [1, 1],
    },
  },
  setup(props, context) {
    // Check win and height
    // eslint-disable-next-line
    if (props.maxWidth && ( props.minWidth > props.maxWidth )) {
      console.warn('[dyno-item warn]: Invalid prop: minWidth cannot be greater than maxWidth');
    }
    // eslint-disable-next-line
    if (props.maxWidth && ( props.minHeight > props.maxHeight )) {
      console.warn('[dyno-item warn]: Invalid prop: minHeight cannot be greater than maxHeight');
    }
    const domRect = reactive({
      bounds: {
        minLeft: null,
        maxLeft: null,
        minRight: null,
        maxRight: null,
        minTop: null,
        maxTop: null,
        minBottom: null,
        maxBottom: null,
      },
      width: null,
      height: null,
      left: props.x,
      top: props.y,
      right: 0,
      bottom: 0,
    });
    const mouseClickPosition = ref(null);
    const root = ref(null);
    const content = ref();
    const eventsFor = ref(events.mouse);
    const parentHeight = ref(null);
    const parentWidth = ref(null);
    const aspectFactor = ref(null);
    const enabled = ref(props.active);

    const resetBoundsAndMouseState = () => {
      mouseClickPosition.value = {
        mouseX: 0, mouseY: 0, x: 0, y: 0, w: 0, h: 0,
      };
      domRect.bounds = {
        minLeft: null,
        maxLeft: null,
        minRight: null,
        maxRight: null,
        minTop: null,
        maxTop: null,
        minBottom: null,
        maxBottom: null,
      };
    };
    resetBoundsAndMouseState();
    const {
      elementMouseDown, elementTouchDown, dragging,
    } = useDrag({
      props,
      context,
      mouseClickPosition,
      domRect,
      eventsFor,
      root,
      enabled,
      parentHeight,
      parentWidth,
      resetBoundsAndMouseState,
    });

    const {
      handleResize, handleDown, handleTouchDown, actualHandles, resizing,
    } = useResize({
      context,
      props,
      domRect,
      mouseClickPosition,
      aspectFactor,
      parentWidth,
      parentHeight,
      eventsFor,
      resetBoundsAndMouseState,
    });
    /**
     * Enable transformations for props x,y and w,h
     */
    useTransform({
      props,
      resizing,
      dragging,
      domRect,
      aspectFactor,
      parentWidth,
      parentHeight,
      content,
    });
    const deselect = (e) => {
      const { target } = e;
      const regex = new RegExp(`${props.className}-([trmbl]{2})`, '');
      if (!root.value.contains(target) && !regex.test(target.className)) {
        if (enabled.value && !props.preventDeactivation) {
          enabled.value = false;
          context.emit('deactivated');
        }
        if (eventsFor.value) {
          removeEvent(document.documentElement, eventsFor.value.move, handleResize);
        }
      }
      resetBoundsAndMouseState();
    };
    const checkParentSize = () => {
      if (props.parent) {
        // TODO understand well
        [parentWidth.value, parentHeight.value] = getParentSize(root.value, props.parent);
        domRect.right = parentWidth.value - domRect.width - domRect.left;
        domRect.bottom = parentHeight.value - domRect.height - domRect.top;
      }
    };
    onMounted(() => {
      if (!props.enableNativeDrag) {
        root.value.ondragstart = () => false;
      }
      [parentWidth.value, parentHeight.value] = getParentSize(root.value, props.parent);
      [domRect.width, domRect.height] = getBoundSize(content.value.children);
      aspectFactor.value = (props.w !== 'auto' ? props.w : domRect.width) / (props.h !== 'auto' ? props.h : domRect.height);
      domRect.width = props.w !== 'auto' ? props.w : domRect.width;
      domRect.height = props.h !== 'auto' ? props.h : domRect.height;
      domRect.right = parentWidth.value - domRect.width - domRect.left;
      domRect.bottom = parentHeight.value - domRect.height - domRect.top;
      if (props.active) {
        context.emit('activated');
      }
      addEvent(document.documentElement, 'mousedown', deselect);
      addEvent(document.documentElement, 'touchend touchcancel', deselect);
      addEvent(window, 'resize', checkParentSize);
    });
    const handleStyle = computed(() => ({
      display: enabled.value ? 'block' : 'none',
    }));
    const style = computed(() => ({
      transform: `translate(${domRect.left}px, ${domRect.top}px)`,
      '-webkit-transition': 'transform 0s ease-in-out',
      width: `${domRect.width}px`,
      height: `${domRect.height}px`,
      zIndex: props.z,
      position: 'absolute',
      ...(dragging.value && props.disableUserSelect ? userSelectNone : userSelectAuto),
    }));
    const contentStyle = computed(() => ({
      width: `${domRect.width}px`,
      height: `${domRect.height}px`,
      position: 'absolute',
      overflow: 'auto',
      ...(dragging.value && props.disableUserSelect ? userSelectNone : userSelectAuto),
    }));
    onBeforeUnmount(() => {
      console.log('before unmount');
      removeEvent(document.documentElement, 'mousedown', deselect);
      removeEvent(document.documentElement, 'touchend touchcancel', deselect);
      removeEvent(window, 'resize', checkParentSize);
    });

    watch(() => props.lockAspectRatio, (val) => {
      if (val) {
        aspectFactor.value = domRect.width / domRect.height;
      } else {
        aspectFactor.value = null;
      }
    });

    watch(() => props.active, (val) => {
      enabled.value = val;
      if (val) {
        context.emit('activated');
      } else {
        context.emit('deactivated');
      }
    });

    return {
      // Dragging feature
      elementMouseDown,
      elementTouchDown,
      dragging,
      // Resize feature
      resizing,
      handleDown,
      handleTouchDown,
      actualHandles,
      // Item
      root,
      style,
      contentStyle,
      handleStyle,
      enabled,
      content,
      aspectFactor,
    };
  },
};
</script>
