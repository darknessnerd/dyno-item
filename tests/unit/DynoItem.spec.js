import { shallowMount } from '@vue/test-utils';
import DynoItem from '@/components/DynoItem.vue';

describe('DynoItem.vue', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(DynoItem);
    expect(wrapper).toMatchSnapshot();
  });
  it('it emit activated onMounted', () => {
    const wrapper = shallowMount(DynoItem, {
      props: { active: true },
    });
    expect(wrapper.emitted().activated[0]).toEqual([]);
    expect(wrapper.vm.style).toStrictEqual({
      transform: 'translate(0px, 0px)',
      '-webkit-transition': 'transform 0s ease-in-out',
      width: '200px',
      height: '200px',
      zIndex: 'auto',
      position: 'absolute',
      userSelect: 'auto',
      MozUserSelect: 'auto',
      WebkitUserSelect: 'auto',
      MsUserSelect: 'auto',
    });
  });
  it('it deselect when mousedown on documentElement', () => {
    const wrapper = shallowMount(DynoItem);
    const event = new window.MouseEvent('mousedown');
    document.documentElement.dispatchEvent(event);
    expect(wrapper.emitted().deactivated[0]).toEqual([]);
  });
});
