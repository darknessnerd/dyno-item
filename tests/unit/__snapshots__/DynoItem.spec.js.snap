// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynoItem.vue renders correctly 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": Object {},
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div
        class="draggable resizable vdr"
        style="transform: translate(0px, 0px); z-index: auto; position: absolute; user-select: auto;"
      >
        
        <div
          class="handle handle-tl"
          data-test="tl"
          style="display: none;"
        >
          
          
        </div>
        <div
          class="handle handle-tm"
          data-test="tm"
          style="display: none;"
        >
          
          
        </div>
        <div
          class="handle handle-tr"
          data-test="tr"
          style="display: none;"
        >
          
          
        </div>
        <div
          class="handle handle-mr"
          data-test="mr"
          style="display: none;"
        >
          
          
        </div>
        <div
          class="handle handle-br"
          data-test="br"
          style="display: none;"
        >
          
          
        </div>
        <div
          class="handle handle-bm"
          data-test="bm"
          style="display: none;"
        >
          
          
        </div>
        <div
          class="handle handle-bl"
          data-test="bl"
          style="display: none;"
        >
          
          
        </div>
        <div
          class="handle handle-ml"
          data-test="ml"
          style="display: none;"
        >
          
          
        </div>
        
        <div
          style="position: absolute; overflow: auto; user-select: auto;"
        >
          
          
        </div>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.5",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "active": false,
    "actualHandles": Array [
      "tl",
      "tm",
      "tr",
      "mr",
      "br",
      "bm",
      "bl",
      "ml",
    ],
    "axis": "both",
    "className": "vdr",
    "classNameActive": "active",
    "classNameDraggable": "draggable",
    "classNameDragging": "dragging",
    "classNameHandle": "handle",
    "classNameResizable": "resizable",
    "classNameResizing": "resizing",
    "content": <div
      style="position: absolute; overflow: auto; user-select: auto;"
    >
      
      
    </div>,
    "contentStyle": Object {
      "MozUserSelect": "auto",
      "MsUserSelect": "auto",
      "WebkitUserSelect": "auto",
      "height": "200px",
      "overflow": "auto",
      "position": "absolute",
      "userSelect": "auto",
      "width": "200px",
    },
    "disableUserSelect": true,
    "dragCancel": null,
    "dragHandle": null,
    "draggable": true,
    "dragging": false,
    "elementMouseDown": [Function],
    "elementTouchDown": [Function],
    "enableNativeDrag": false,
    "enabled": false,
    "grid": Array [
      1,
      1,
    ],
    "h": 200,
    "handleDown": [Function],
    "handleTouchDown": [Function],
    "handles": Array [
      "tl",
      "tm",
      "tr",
      "mr",
      "br",
      "bm",
      "bl",
      "ml",
    ],
    "lockAspectRatio": false,
    "maxHeight": null,
    "maxWidth": null,
    "minHeight": 0,
    "minWidth": 0,
    "onDrag": [Function],
    "onDragStart": [Function],
    "onResize": [Function],
    "onResizeStart": [Function],
    "parent": false,
    "preventDeactivation": false,
    "resizable": true,
    "resizing": false,
    "root": <div
      class="draggable resizable vdr"
      style="transform: translate(0px, 0px); z-index: auto; position: absolute; user-select: auto;"
    >
      
      <div
        class="handle handle-tl"
        data-test="tl"
        style="display: none;"
      >
        
        
      </div>
      <div
        class="handle handle-tm"
        data-test="tm"
        style="display: none;"
      >
        
        
      </div>
      <div
        class="handle handle-tr"
        data-test="tr"
        style="display: none;"
      >
        
        
      </div>
      <div
        class="handle handle-mr"
        data-test="mr"
        style="display: none;"
      >
        
        
      </div>
      <div
        class="handle handle-br"
        data-test="br"
        style="display: none;"
      >
        
        
      </div>
      <div
        class="handle handle-bm"
        data-test="bm"
        style="display: none;"
      >
        
        
      </div>
      <div
        class="handle handle-bl"
        data-test="bl"
        style="display: none;"
      >
        
        
      </div>
      <div
        class="handle handle-ml"
        data-test="ml"
        style="display: none;"
      >
        
        
      </div>
      
      <div
        style="position: absolute; overflow: auto; user-select: auto;"
      >
        
        
      </div>
    </div>,
    "scale": 1,
    "style": Object {
      "-webkit-transition": "transform 0s ease-in-out",
      "MozUserSelect": "auto",
      "MsUserSelect": "auto",
      "WebkitUserSelect": "auto",
      "height": "200px",
      "position": "absolute",
      "transform": "translate(0px, 0px)",
      "userSelect": "auto",
      "width": "200px",
      "zIndex": "auto",
    },
    "w": 200,
    "x": 0,
    "y": 0,
    "z": "auto",
  },
  "rootVM": Object {},
}
`;
